tls/s2n_security_policies.c

```diff
@@ -1459,3 +1459,19 @@ S2N_RESULT s2n_security_policy_get_version(const struct s2n_security_policy *sec
     }
     RESULT_BAIL(S2N_ERR_INVALID_SECURITY_POLICY);
 }
+
+S2N_RESULT s2n_security_policy_validate_sig_scheme_supported(
+        const struct s2n_signature_preferences *cert_sig_preferences,
```

Comment:
Why does an s2n_security_policy_* method take an s2n_signature_preferences struct? This should probably either take an s2n_security_policy struct or be s2n_signature_preferences_validate_supported.

The comment in the header file is also a little odd, since we don't know that the signature_preferences you pass in are the certificate_signature_preferences list. It could take any list.

Reply:
Updated! I had always treated the prefixes as a sort of rough namespace. E.g. "this validation method has to do with security policies, so it's in the `s2n_security_policy` namespace. Is there any guidance on when things should be exactly prefixed (`s2n_signature_preferences_validate_supported`) vs a more general "namespace" approach like [this](https://github.com/aws/s2n-tls/pull/4379/files#r1479108243)

Reply:
If you were writing this in rust, you probably wouldn't define s2n_security_policy::validate_sig_scheme_supported(cert_sig_preferences, info). You'd probably define s2n_signature_scheme::validate_sig_scheme_supported(self, info). I think it's worth considering how you'd do this in a language with more formal organization :)

The prefixes are a rough namespace, and that /usually/ also applies to the struct they operate on. That's not always true, but I'd say when it's not true the method is either operating on s2n_connection (we can't just stick _everything_ in s2n_connection :P), there at least isn't a different prefix that makes more logical sense, or the method is static. But getting lazy with the static method names can bite you if they ever become not static :(

Often if the method operates on something other than its namespace struct, the name should reflect that-- something like s2n_security_policy_cert_signature_scheme_validate? In particular, if the namespace prefix makes the function name deceptive (an s2n_security_policy isn't actually validating anything in your method) then that's a strong sign you've either got the wrong module or need to otherwise rethink your name.

Anyway, you now have an s2n_signature_preferences method in s2n_security_policies, but the name and what it operates on suggests it belongs in a s2n_signature_preferences module. It looks like that module doesn't exist though-- s2n_signature_preferences is defined in [s2n_signature_scheme](https://github.com/aws/s2n-tls/blob/6c7cdc8329d846d7fcc8e505cd7708e3f2604d70/tls/s2n_signature_scheme.h#L37) instead. So I'd say that this method probably belongs in s2n_signature_scheme. Some of the older code is a bit disorganized :/ Or if you really want to keep it in s2n_security_policies, figure out a name that doesn't indicate that it operates on a security policy.

Reply:
talked offline. Recap
1. generally my first understanding of namespace as prefix was correct
2. and my first intention of putting it in the s2n_security_policy namespace was also probably the right move
3. so it should take an s2n_security_policy